scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala

```diff
@@ -510,14 +511,52 @@ sealed trait SCollection[T] extends PCollectionWrapper[T] {
   def groupBy[K](
     f: T => K
   )(implicit kcoder: Coder[K], vcoder: Coder[T]): SCollection[(K, Iterable[T])] =
+    groupMap(f)(identity)
+
+  /**
+   * Return an SCollection of grouped items. Each group consists of a key and a sequence of
+   * elements transformed into a value of type `U`. The ordering of elements within each group is not guaranteed,
+   * and may even differ each time the resulting SCollection is evaluated.
+   *
+   * It is equivalent to groupBy(key).mapValues(_.map(f)), but more efficient.
+   *
+   * @group transform
+   */
+  def groupMap[K, U](f: T => K)(
+    g: T => U
+  )(implicit kc: Coder[K], uc: Coder[U]): SCollection[(K, Iterable[U])] =
     this.transform {
-      val coder = CoderMaterializer.kvCoder[K, T](context)
-      _.pApply(WithKeys.of(Functions.serializableFn(f)))
-        .setCoder(coder)
-        .pApply(GroupByKey.create[K, T]())
+      val cf = ClosureCleaner.clean(f)
+      val cg = ClosureCleaner.clean(g)
+
+      _.map(t => KV.of(cf(t), cg(t)))
+        .setCoder(CoderMaterializer.kvCoder[K, U](context))
+        .pApply(GroupByKey.create[K, U]())
         .map(kvIterableToTuple)
     }
 
+  /**
+   * Return an SCollection of grouped items. Each group consists of a key and the result of an associative
+   * reduce function. The ordering of elements within each group is not guaranteed, and may even differ each
+   * time the resulting SCollection is evaluated.
+   *
+   * The associative function is performed locally on each mapper before sending results to
+   * a reducer, similarly to a "combiner" in MapReduce
+   *
+   * @group transform
+   */
+  def groupMapReduce[K](f: T => K)(
```

Comment:
How does this compare to the more idiomatic `keyBy(f),reduceByKey(g)`?

Reply:
Well good question, there's a tiny bit advantage over that example, where there's one step less of conversion of `tuple` to `kv`. Also be because we wrap this in a composite transform might increase chances of fusion.

Recently also scala 2.13 added these methods to collections, so i guess this might become idiomatic?